File: 백준/Silver/1926. 그림/그림.java
import java.util.*;
import java.io.*;



public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int[] input = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int n = input[0];
        int m = input[1];

        int[][] map = new int[n][m];
        boolean[][] visited = new boolean[n][m];


        for (int i = 0; i < n; ++i) {
            map[i] = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        }

        int count = 0;
        int max_size = 0;
        int[] dy = {-1, 1, 0, 0};
        int[] dx = {0, 0, -1, 1};

        for (int y = 0; y < n; ++y) {
            for (int x = 0; x < m; ++x) {
                if (map[y][x] == 0) continue;
                if (visited[y][x]) continue;
                ++count;
                int size = 0;
                Queue<int[]> q = new LinkedList<>();
                q.offer(new int[]{y, x});
                visited[y][x] = true;
                ++size;
                while(!q.isEmpty()) {
                    int[] cur = q.poll();
                    for (int dir = 0; dir < 4; ++dir) {
                        int ny = cur[0] + dy[dir];
                        int nx = cur[1] + dx[dir];
                        if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue;
                        if (map[ny][nx] == 0) continue;
                        if (visited[ny][nx]) continue;
                        visited[ny][nx] = true;
                        ++size;
                        q.add(new int[]{ny, nx});
                    }
                }
                if (max_size < size) max_size = size;
            }
        }

        bw.write(count + "\n");
        bw.write( max_size + "\n");

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/1926. 그림/그림.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.`out`))

fun main() {
    val size = br.readLine().split(' ').map(String::toInt).toIntArray()
    val n = size[0]
    val m = size[1]
    val paper = Array(n) { IntArray(m) {0} }
    val visited = Array(n) { BooleanArray(m) {false} }

    for (y in 0..<n) {
        val line = br.readLine().split(' ').map(String::toInt).toIntArray()
        for (x in line.indices) {
            paper[y][x] = line[x]
        }
    }

    var dy = arrayOf(0, 0, -1, 1)
    var dx = arrayOf(-1, 1, 0, 0)

    var count = 0
    var max = 0

    for (r in 0 ..<n) {
        for (c in 0 ..<m) {
            if (paper[r][c] == 0) continue
            if (visited[r][c]) continue

            val q = LinkedList<Pair<Int, Int>>()
            ++count
            var size = 1
            visited[r][c] = true
            q.add(Pair(r, c))

            while(q.isNotEmpty()) {
                val (y, x) = q.removeFirst()

                for (dir in 0..<4) {
                    val ny = y + dy[dir]
                    val nx = x + dx[dir]

                    if (ny < 0 || ny >= n || nx < 0 || nx >= m) continue
                    if (paper[ny][nx] == 0) continue
                    if (visited[ny][nx]) continue

                    ++size
                    visited[ny][nx] = true
                    q.add(Pair(ny, nx))
                }
            }
            max = max.coerceAtLeast(size)
        }
    }
    
    bw.write("$count\n$max")

    bw.flush()
    bw.close()
    br.close()
}


